followthemoney/statement/statement.py

```diff
@@ -0,0 +1,239 @@
+                RuntimeWarning,
+            )
+        return hash(self.id)
+
+    def __repr__(self) -> str:
+        return "<Statement(%r, %r, %r)>" % (self.entity_id, self.prop, self.value)
+
+    def __eq__(self, other: Any) -> bool:
+        return not self.id != other.id
+
+    def __lt__(self, other: Any) -> bool:
+        self_key = (self.prop != BASE_ID, self.id or "")
+        other_key = (other.prop != BASE_ID, other.id or "")
+        return self_key < other_key
+
+    def clone(self: Self) -> "Statement":
+        """Make a deep copy of the given statement."""
+        return Statement.from_dict(self.to_dict())
+
+    def generate_key(self) -> Optional[str]:
+        return self.make_key(
+            self.dataset,
+            self.entity_id,
+            self.prop,
+            self.value,
+            self.external,
+        )
+
+    @classmethod
+    def make_key(
+        cls,
+        dataset: str,
+        entity_id: str,
+        prop: str,
+        value: str,
+        external: Optional[bool],
+    ) -> Optional[str]:
+        """Hash the key properties of a statement record to make a unique ID."""
+        if prop is None or value is None:
+            return None
+        key = f"{dataset}.{entity_id}.{prop}.{value}"
+        if external:
+            # We consider the external flag in key composition to avoid race conditions
+            # where a certain entity might be emitted as external while it is already
+            # linked in to the graph via another route.
+            key = f"{key}.ext"
+        return hashlib.sha1(key.encode("utf-8")).hexdigest()
+
+    @classmethod
+    def from_dict(cls, data: StatementDict) -> "Statement":
+        return cls(
+            entity_id=data["entity_id"],
+            prop=data["prop"],
+            schema=data["schema"],
+            value=data["value"],
+            dataset=data["dataset"],
+            lang=data.get("lang", None),
+            original_value=data.get("original_value", None),
+            first_seen=data.get("first_seen", None),
+            external=data.get("external", False),
+            id=data.get("id", None),
+            canonical_id=data.get("canonical_id", None),
+            last_seen=data.get("last_seen", None),
+        )
+
+    @classmethod
+    def from_db_row(cls, row: Row) -> "Statement":
+        return cls(
+            id=row.id,
+            canonical_id=row.canonical_id,
+            entity_id=row.entity_id,
+            prop=row.prop,
+            schema=row.schema,
+            value=row.value,
+            dataset=row.dataset,
+            lang=row.lang,
+            original_value=row.original_value,
+            first_seen=datetime_iso(row.first_seen),
+            external=row.external,
+            last_seen=datetime_iso(row.last_seen),
+        )
+
+    @classmethod
+    def from_entity(
+        cls,
+        entity: "StatementEntity",
+        dataset: str,
```

Comment:
I notice that this works perfectly well with the old fashioned `EntityProxy`. Is there an issue with expanding this signature to include that? Seems like a great compatibility feature.
```suggestion
        entity: "EntityProxy",
```

Reply:
Example:

```
person: EntityProxy = model.make_entity("Person")
person.make_id("test_person")
person.add("name", "John Doe")
person.add("birthDate", "1980-01-01")
person.add("email", "john.doe@example.com")

for statement in Statement.from_entity(person, dataset="test"):
    print(statement.to_dict())

=====OUTPUT=====
{'canonical_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'entity_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'prop': 'id', 'schema': 'Person', 'value': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'dataset': 'test', 'lang': None, 'original_value': None, 'first_seen': None, 'last_seen': None, 'external': False, 'id': '9385d7840a5a553d58a5a5e10491b9d56cdc35e8'}
{'canonical_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'entity_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'prop': 'name', 'schema': 'Person', 'value': 'John Doe', 'dataset': 'test', 'lang': None, 'original_value': None, 'first_seen': None, 'last_seen': None, 'external': False, 'id': 'ab5a7ceab0ea9342546f2acaa18d9e9605328303'}
{'canonical_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'entity_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'prop': 'birthDate', 'schema': 'Person', 'value': '1980-01-01', 'dataset': 'test', 'lang': None, 'original_value': None, 'first_seen': None, 'last_seen': None, 'external': False, 'id': 'c7afdd693fe54b30ae043a962b84ba1e10b94fe9'}
{'canonical_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'entity_id': 'f73e8656f8a7a531b33d474fc54f9759e3343bf5', 'prop': 'email', 'schema': 'Person', 'value': 'john.doe@example.com', 'dataset': 'test', 'lang': None, 'original_value': None, 'first_seen': None, 'last_seen': None, 'external': False, 'id': '66165c4fbf27beb339b72141669cce7a4edddede'}
```
No data is added, but the format is compatible.

Reply:
Actually, it doesn't seem to work in the expected way for the `StatementsEntity`, because the members of `Statement` are set without regard for the input values.

e.g.
```
    person: StatementEntity = StatementEntity(
        Dataset.make({"name": "testdataset"}),
        {"schema": "Person", "properties": {}},
    )
    person.id = "test_id"
    person.add_statement(
        Statement(
            entity_id="test_id",
            prop="education",
            schema="Person",
            value="University of California, Los Angeles",
            dataset="testdataset",
            first_seen="2025-01-01",
            last_seen="2025-01-01",
        ),
    )
    person.add_statement(
        Statement(
            entity_id="test_id",
            prop="education",
            schema="Person",
            value="University of California, Los Angeles",
            dataset="testdataset",
            last_seen="2025-01-02",
        ),
    )
    print("--------------------------------")
    for s in Statement.from_entity(person, dataset="testdataset"):
        print(s.to_db_row())
```
returns
```
{
    'canonical_id': 'test_id',
    'entity_id': 'test_id',
    'prop': 'id',
    'schema': 'Person',
    'value': 'test_id',
    'dataset': 'testdataset',
    'lang': None,
    'original_value': None,
    'first_seen': None,
    'last_seen': None,
    'external': False,
    'id': '54cbed65f3a51c2c859014bebd91529f0a77557b'
}
{
    'canonical_id': 'test_id',
    'entity_id': 'test_id',
    'prop': 'education',
    'schema': 'Person',
    'value': 'University of California, Los Angeles',
    'dataset': 'testdataset',
    'lang': None,
    'original_value': None,
    'first_seen': None,
    'last_seen': None,
    'external': False,
    'id': '2509f49fd78932551c199f1b9f47a91aafa40b22'
}
```
So none of the data from the Statements other than the value has propagated. Am I missing something here @pudo?

Reply:
@pudo 
I can modify this to use EntityProxy instead, which basically 'upgrades' an FTM <4 statement to 4.0, is that the intended usecase? 
Or it can use StatementEntity and in that case, I feel like the statement data should propagate down, unless I'm missing something? Just lmk